src/Mod/Part/Gui/ViewProviderGridExtension.cpp

```diff
@@ -0,0 +1,513 @@
+        return;
+    }
+
+    int numberOfLines = static_cast<int>(std::max(pixelWidth, pixelHeight)) / GridSizePixelThreshold;
+
+    double unitMultiplier = (unitsUserSchema == 2 || unitsUserSchema == 3) ? 25.4 : (unitsUserSchema == 5 || unitsUserSchema == 7) ? 304.8 : 1;
+
+    computedGridValue = vp->GridSize.getValue() * unitMultiplier * pow(GridNumberSubdivision, floor(log(camMaxDimension / unitMultiplier / numberOfLines) / log(GridNumberSubdivision)));
+
+    //cap the grid size
+    computedGridValue = std::max(computedGridValue, 0.000001);
+    computedGridValue = std::min(computedGridValue, 10000000.0);
+}
+
+void GridExtensionP::createGrid(bool cameraUpdate)
+{
+    Gui::MDIView* mdi = Gui::Application::Instance->editDocument()->getActiveView();
+    Gui::View3DInventorViewer* viewer = static_cast<Gui::View3DInventor*>(mdi)->getViewer();
+
+    bool cameraDimensionsChanged = checkCameraZoomChange(viewer);
+
+    bool cameraCenterMoved = checkCameraTranslationChange(viewer);
+
+    bool gridNeedUpdating = cameraDimensionsChanged || cameraCenterMoved;
+
+    if (!gridNeedUpdating && cameraUpdate)
+        return;
+
+    Gui::coinRemoveAllChildren(GridRoot);
+
+    computeGridSize(viewer);
+
+    auto getColor = [](auto unpackedcolor) {
+        SoBaseColor* lineColor = new SoBaseColor;
+        float transparency;
+        SbColor lineCol(0.7f, 0.7f, 0.7f);
+        lineCol.setPackedValue(unpackedcolor, transparency);
+        lineColor->rgb.setValue(lineCol);
+        return lineColor;
+    };
+
+    //First we create the subdivision lines
+    createGridPart(GridNumberSubdivision, true,
+                   (GridNumberSubdivision == 1), GridLinePattern,
+                   getColor(GridLineColor), GridLineWidth);
+
+    //Second we create the wider lines marking every nth lines
+    if (GridNumberSubdivision > 1) {
+        createGridPart(GridNumberSubdivision, false, true,
+            GridDivLinePattern, getColor(GridDivLineColor), GridDivLineWidth);
+    }
+}
+
+void GridExtensionP::createGridPart(int numberSubdiv, bool subDivLines, bool divLines, int pattern, SoBaseColor* color, int lineWidth)
+{
+    SoGroup* parent = new Gui::SoSkipBoundingGroup();
+    GridRoot->addChild(parent);
+    SoVertexProperty* vts;
+
+    parent->addChild(color);
+
+    SoDrawStyle* DefaultStyle = new SoDrawStyle;
+    DefaultStyle->lineWidth = lineWidth;
+    DefaultStyle->linePattern = pattern;
+    parent->addChild(DefaultStyle);
+
+    SoPickStyle* PickStyle = new SoPickStyle;
+    PickStyle->style = SoPickStyle::UNPICKABLE;
+    parent->addChild(PickStyle);
+
+    SoLineSet* grid = new SoLineSet;
+    vts = new SoVertexProperty;
+    grid->vertexProperty = vts;
+
+    float gridDimension = 1.5 * camMaxDimension;
+    int vlines = static_cast<int>(gridDimension / computedGridValue); // total number of vertical lines
+    int nlines = 2 * vlines; // total number of lines
+    if (nlines > 2000) {
+        Gui::coinRemoveAllChildren(GridRoot);
```

Comment:
Maybe would be worth print a warning or something in this situation ?

Reply:
I don't think it's necessary, it feels normal. I think after you try it you'll see what I mean. A warning would just bloat the user.

Reply:
I imagine what it does. :) Or say a log message (eventually only in debug builds).

Reply:
Here the problem I see is that this function maybe called quite often and quite a loot.

So, should the user be informed that the grid is not shown because it would be useless given the amount of segments, so that he or she can adjust the settings of the grid?  I think it makes sense.

Does it make sense only in Debug mode? I do not think it does. If the developer is looking into the grid, he will understand pretty fast. If not, then most probably the developer does not care at all.

So, either we provide a mechanism where a single notification is sent or we leave it like this.

I am going with providing a notification. Ideally, this kind of notification should be handled with the new notification framework as a user notification (it is neither a warning, nor an error, nor a log), but it is not merged yet, so it goes as a warning (so that it annoys enough and we remember to change it to a user notification when that is merged).